File: go/bob/bob.go
package bob

import (
	"strings"
	"unicode"
)

const testVersion = 3

// Hey return Bob's response to Alice's question
func Hey(in string) string {
	in = strings.TrimSpace(in)
	if isSilence(in) {
		return "Fine. Be that way!"
	}
	if hasLetters(in) && isYelling(in) {
		return "Whoa, chill out!"
	}
	if isQuestion(in) {
		return "Sure."
	}
	return "Whatever."
}

func hasLetters(s string) bool {
	for _, c := range s {
		if unicode.IsLetter(c) {
			return true
		}
	}
	return false
}

func isYelling(s string) bool {
	for _, c := range s {
		if unicode.IsLower(c) {
			return false
		}
	}
	return true
}

func isQuestion(s string) bool {
	if s[len(s)-1:] == "?" {
		return true
	}
	return false
}

func isSilence(s string) bool {
	for _, c := range s {
		if !unicode.IsSpace(c) {
			return false
		}
	}
	return true
}

File: csharp/bob/Bob.cs
using System;

public static class Bob {
    private const string silence = "Fine. Be that way!",
        yelling = "Whoa, chill out!",
        question = "Sure.",
        whatever = "Whatever.";

    public static string Response (string statement) {
        statement = statement.Trim ();
        if (isSilence (statement)) {
            return silence;
        }
        if (isYelling (statement)) {
            return yelling;
        }
        if (isQuestion (statement)) {
            return question;
        }
        return whatever;
    }

    private static bool isSilence (string statement) {
        if (string.IsNullOrWhiteSpace (statement)) {
            return true;
        }
        return false;
    }

    private static bool isYelling (string statement) {
        bool foundUpper = false;
        foreach (char c in statement) {
            if (char.IsLower (c)) {
                return false;
            }
            if (char.IsUpper (c)) {
                foundUpper = true;
            }
        }
        return foundUpper;
    }
    private static bool isQuestion (string statement) {
        if (statement.EndsWith ("?")) {
            return true;
        }
        return false;
    }
}